### DISCUSSION TITLE
Support kafka client 3.4.1 in spring-kafka-test

### DISCUSSION THREAD

**Discussion by tomlicha** [spring-projects/spring-kafka#2752] [General]
Hi,

I've recently bumped into an issue trying to upgrade to kafka client 3.4.1 with spring boot 3.1.1, it appears that the spring-kafka-test dependency still uses kafka-client 3.3.2. Is there an upgrade already scheduled ?

Best

**Comment by garyrussell** [comment]
Spring versioning rules do not allow us to change minor versions of dependencies in a patch release.

The documentation shows how to override the dependency (3.5.1 is now available).

See https://docs.spring.io/spring-kafka/docs/current/reference/html/#update-deps

When upgrading one version, you must upgrade them all; when you use Boot's dependency management, that will happen automatically, because Boot uses `spring-kafka-bom`.

**Comment by tomlicha** [comment]
Hi, I have upgraded to 3.5.1, but in EmbeddedKafkaBroker they still call test utils from kafka_2.13:3.3.2, is there a workaround ? I cannot just exclude 3.3.2 because otherwise, test utils is no more defined

**Reply by garyrussell** [reply]
You have to override all the kafka jar versions to 35.1. If you use Spring Boot; its dependency management will take care of it by just setting `spring-kafka.version` and `kafka.version`. If you are not using Spring Boot's dependency management; you must override all the kafka jars yourself to matching versions. Show your pom.

**Reply by tomlicha** [reply]
I am using :
`<kafka.version>3.5.1</kafka.version>`
and 
`<dependency>
                <groupId>org.springframework.kafka</groupId>
                <artifactId>spring-kafka</artifactId>
                <version>${springframework-kafka.version}</version>
            </dependency>
            <dependency>
                <groupId>org.springframework.kafka</groupId>
                <artifactId>spring-kafka-test</artifactId>
                <version>${springframework-kafka.version}</version>
                <scope>test</scope>
            </dependency>
            <dependency>
                <groupId>org.apache.kafka</groupId>
                <artifactId>kafka_2.13</artifactId>
                <version>${kafka.version}</version>
            </dependency>`

**Reply by garyrussell** [reply]
If you use Boot's dependency management; you don't need a custom `version` for spring-kafka, just set `spring-kafka.version` and leave the version off; same thing for `kafka.version`. Boot will override all the jars according to the Kafka BOM.

**Reply by tomlicha** [reply]
so you mean I shoudl remove :
`      
      <dependency>
                <groupId>org.springframework.kafka</groupId>
                <artifactId>spring-kafka</artifactId>
                <version>${springframework-kafka.version}</version>
            </dependency>
`

**Reply by garyrussell** [reply]
It depends on whether you are using Boot's dependency management; if you are, just remove the `<version/>` element (and set `spring-kafka.version`).

If you are not using Boot's dependency management; you will have to override the versions of the kafka test classified jars. Older versions of Boot did not manage those jars (it does now, so that implies you are not using dependency management).

The older docs show how to set those dependencies:

https://docs.spring.io/spring-kafka/docs/2.8.11/reference/html/#update-deps

```xml
<dependency>
    <groupId>org.apache.kafka</groupId>
    <artifactId>kafka-clients</artifactId>
    <classifier>test</classifier>
    <scope>test</scope>
    <version>${kafka.version}</version>
</dependency>

<dependency>
    <groupId>org.apache.kafka</groupId>
    <artifactId>kafka_2.13</artifactId>
    <classifier>test</classifier>
    <scope>test</scope>
    <version>${kafka.version}</version>
</dependency>
```

**Reply by garyrussell** [reply]
`./mvnw :dependency:tree` will show you what's wrong.

**Reply by garyrussell** [reply]
See 

https://docs.spring.io/spring-boot/docs/current/reference/html/using.html#using.build-systems.dependency-management

and 

https://docs.spring.io/spring-boot/docs/3.1.3/maven-plugin/reference/htmlsingle/

about using dependency management.

**Reply by garyrussell** [reply]
It makes it much easier to override versions, as we show here:

https://docs.spring.io/spring-kafka/docs/current/reference/html/#update-deps

**Reply by tomlicha** [reply]
appologies, I've sent you the parent pom, here is what i am using at the test level : 

```xml
        <dependency>
            <groupId>org.apache.kafka</groupId>
            <artifactId>kafka_2.13</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.kafka</groupId>
            <artifactId>spring-kafka</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.kafka</groupId>
            <artifactId>spring-kafka-test</artifactId>
            <scope>test</scope>
        </dependency>
```

**Reply by garyrussell** [reply]
Please learn GitHub markdown (see after my edit).

**Reply by garyrussell** [reply]
This is the pom from my smoke test, that overrides the `kafka.version` in a Boot 3.0.10/3.1.3 project...

```xml
	<properties>
		<java.version>17</java.version>
		<kafka.version>3.5.1</kafka.version>
	</properties>
	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.kafka</groupId>
			<artifactId>spring-kafka</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>org.springframework.kafka</groupId>
			<artifactId>spring-kafka-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>
```
